src/main/java/org/opensearch/security/ssl/SslContextHandler.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.ssl;
+
+import java.nio.charset.StandardCharsets;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.net.ssl.SSLEngine;
+
+import org.opensearch.security.ssl.config.Certificate;
+import org.opensearch.transport.NettyAllocator;
+
+import io.netty.handler.ssl.SslContext;
+
+public class SslContextHandler {
+
+    private SslContext sslContext;
+
+    private final SslConfiguration sslConfiguration;
+
+    private final List<Certificate> loadedCertificates;
+
+    public SslContextHandler(final SslConfiguration sslConfiguration) {
+        this(sslConfiguration, false);
+    }
+
+    public SslContextHandler(final SslConfiguration sslConfiguration, final boolean client) {
+        this.sslContext = client ? sslConfiguration.buildClientSslContext(true) : sslConfiguration.buildServerSslContext(true);
+        this.sslConfiguration = sslConfiguration;
+        this.loadedCertificates = sslConfiguration.certificates();
+    }
+
+    public SSLEngine createSSLEngine() {
+        return sslContext.newEngine(NettyAllocator.getAllocator());
+    }
+
+    public SSLEngine createSSLEngine(final String hostname, final int port) {
+        return sslContext.newEngine(NettyAllocator.getAllocator(), hostname, port);
+    }
+
+    public SslConfiguration sslConfiguration() {
+        return sslConfiguration;
+    }
+
+    SslContext sslContext() {
+        return sslContext;
+    }
+
+    public Stream<Certificate> keyMaterialCertificates() {
+        return keyMaterialCertificates(loadedCertificates);
+    }
+
+    Stream<Certificate> keyMaterialCertificates(final List<Certificate> certificates) {
+        return certificates.stream().filter(Certificate::hasKey);
+    }
+
+    void reloadSslContext() throws CertificateException {
+        final var newCertificates = sslConfiguration.certificates();
+
+        if (sameCertificates(newCertificates)) {
+            return;
+        }
+
+        validateNewCertificates(newCertificates);
+        invalidateSessions();
```

Comment:
this will change the behavior for sessions w.r.t. TLS handling of the plugin today. Can this create a problem for large clusters where certificates are being updated in a rolling-restart manner and each node needs to support both old & new certificate sessions?

Reply:
No it is not. JDK supports session resumption. From official docs for invalidate method:
```
Invalidates the session.
Future connections will not be able to resume or join this session. However, any existing connection using this session can continue to use the session until the connection is closed.
```